source/opt/dead_branch_elim_pass.cpp

```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
-  auto ii = bp->end();
-  --ii;
-  *branchInst = &*ii;
-  if (ii == bp->begin()) return false;
-  --ii;
-  *mergeInst = &*ii;
-  if ((*mergeInst)->opcode() != SpvOpSelectionMerge) return false;
-  // SPIR-V says the terminator for an OpSelectionMerge must be
-  // either a conditional branch or a switch.
-  assert((*branchInst)->opcode() == SpvOpBranchConditional ||
-         (*branchInst)->opcode() == SpvOpSwitch);
-  // Both BranchConidtional and Switch have their conditional value at 0.
-  *condId = (*branchInst)->GetSingleWordInOperand(0);
-  return true;
-}
-
-bool DeadBranchElimPass::HasNonPhiNonBackedgeRef(uint32_t labelId) {
-  bool nonPhiNonBackedgeRef = false;
-  get_def_use_mgr()->ForEachUser(
-      labelId, [this, &nonPhiNonBackedgeRef](ir::Instruction* user) {
-        if (user->opcode() != SpvOpPhi &&
-            backedges_.find(user) == backedges_.end()) {
-          nonPhiNonBackedgeRef = true;
-        }
-      });
-  return nonPhiNonBackedgeRef;
-}
-
-void DeadBranchElimPass::ComputeBackEdges(
-    std::list<ir::BasicBlock*>& structuredOrder) {
-  backedges_.clear();
-  std::unordered_set<uint32_t> visited;
-  // In structured order, edges to visited blocks are back edges
-  for (auto bi = structuredOrder.begin(); bi != structuredOrder.end(); ++bi) {
-    visited.insert((*bi)->id());
-    auto ii = (*bi)->end();
-    --ii;
-    switch (ii->opcode()) {
-      case SpvOpBranch: {
-        const uint32_t labId =
-            ii->GetSingleWordInOperand(kBranchTargetLabIdInIdx);
-        if (visited.find(labId) != visited.end()) backedges_.insert(&*ii);
-      } break;
-      case SpvOpBranchConditional: {
-        const uint32_t tLabId =
-            ii->GetSingleWordInOperand(kBranchCondTrueLabIdInIdx);
-        if (visited.find(tLabId) != visited.end()) {
-          backedges_.insert(&*ii);
-          break;
-        }
-        const uint32_t fLabId =
-            ii->GetSingleWordInOperand(kBranchCondFalseLabIdInIdx);
-        if (visited.find(fLabId) != visited.end()) backedges_.insert(&*ii);
-      } break;
-      default:
-        break;
-    }
-  }
-}
-
 bool DeadBranchElimPass::EliminateDeadBranches(ir::Function* func) {
-  // Traverse blocks in structured order
-  std::list<ir::BasicBlock*> structuredOrder;
-  cfg()->ComputeStructuredOrder(func, &*func->begin(), &structuredOrder);
-  ComputeBackEdges(structuredOrder);
-  std::unordered_set<ir::BasicBlock*> elimBlocks;
-  bool modified = false;
-  for (auto bi = structuredOrder.begin(); bi != structuredOrder.end(); ++bi) {
-    // Skip blocks that are already in the elimination set
-    if (elimBlocks.find(*bi) != elimBlocks.end()) continue;
-    // Skip blocks that don't have conditional branch preceded
-    // by OpSelectionMerge
-    ir::Instruction* br;
-    ir::Instruction* mergeInst;
-    uint32_t condId;
-    if (!GetSelectionBranch(*bi, &br, &mergeInst, &condId)) continue;
+  auto get_parent_block = [this](uint32_t id) {
```

Comment:
Should say that we only need the instr-to-block mapping in for instructions that exist when we start processing or that we explicitly analyze later. You don't need to update it for instructions that are removed.
That explains why later creation of branch instructions don't need to invalidate or update the analysis.

Reply:
I'll add a comment.